Fix for pickling a Network

dkeeney opened the pull request:
This PR fixes the pickling code in the python interface.

dkeeney commented:
This is not working yet.  I will be gone over the weekend.  Thought that if anyone else would like to work on this they are free to do so.
The current problem is when a network is pickled, a python implemented region is not being pickled correctly.

`Python --> C++ Network class --> C++ PyBiindRegion --> Python implemented region`

The unit test `testNetworkPickle` in bindings/py/tests/regions/network_test.py shows the problem.
I have debug statements scattered around so you can see it has an exception thrown at line 369 of PyBindRegion.cpp while trying to create an in-memory stream that can be used by the pickle.load function.

Enjoy.

dkeeney commented:
I don't see that anyone took up the challenge of trying to fix this so I will continue where I left off from last week.

dkeeney commented:
This addresses issue #732

dkeeney commented:
Pickling of entire Network is now working.
@breznak , @fcr please review when you find time.

breznak commented on bindings/py/cpp_src/bindings/engine/py_Engine.cpp:
```diff
@@ -457,7 +457,7 @@ namespace htm_ext
                 Network self;
                 self.load(ss);
-                return std::move(self);  
+                return self;  
```
out of curiosity, why was this the problem?

dkeeney replied:
> out of curiosity, why was this the problem?

The `std::move(self) ` code worked fine on MSVC but when I submitted it, the OSx build told me I cannot return a local object that way and it said tor remove the std::move( ).  Since the Network class has a move constructor and no copy constructor it automatically does a move when returning by value.

fcr commented:
Thank you @dkeeney.

I built the branch. The **pickle** works without an error  but I get an when trying to **unpickle** any of the location networks in `htm.advanced.frameworks.locatation.location_network_creation.py`:

The error is:
`RuntimeError: CHECK FAILED: "srcOutput.get() != o" addLink -- link from region L4 output winnerCells to region L6a input anchorGrowthCandidates already exists`

I have attached the test file: 
[L46aNetwork_test.py.zip](https://github.com/htm-community/htm.core/files/3814103/L46aNetwork_test.py.zip)

dkeeney commented:
>  Could you or @fcr please add the reproducible test for the error reported?

Fred did provide python unit tests and I added them.  I also created a C++ version of the same test and added that as well.
- htm.core/bindings/py/tests/regions/network_test.py  `def testNetworkPickle(self):`
- htm.core/bindings/py/tests/regions/pyregion_test.py `def testPickle(self):`
- htm.core/src/test/unit/engine/NetworkTest.cpp  `TEST(NetworkTest, SaveRestore) {`
- I also un-skipped the other unit tests in network_test.py (they were pickle related).

breznak commented on src/htm/engine/Network.cpp:
```diff
@@ -44,6 +44,16 @@ Network::Network() {
 
+// move constructor
+Network::Network(Network && n) {
+  regions_ = std::move(n.regions_);
```
Network(API) classes do not use the concept of `initialize()` method for Cereal to work? (needs implicit constructor + initialize() method)

dkeeney replied:
> (needs implicit constructor + initialize() method)

When Cereal restores it assumes that the original saved stream had been initialized before storing.  So, it tries to put everything back to where it was.

I found that if the original network had not been initialized before being serialized, the restore has a problem because some of the things it expected were not there.  I am looking into this and a few other serialization problems right now (to be in a follow-on PR).

The py/advanced unit tests have some non-trivial tests and it has exposed a few new problems I did not know we had. :-)

But to answer what I think is your question: 
- Before serialization we do constructor followed by adding regions and links, then initialize.  The initialize step will be performed by run( ) if needed.
- On restore we start with an empty constructor and then populate everything so we do not need the initialization step.
